src/mongo_odm/query_builder.hpp

```diff
@@ -0,0 +1,473 @@
+// Copyright 2016 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <mongo_odm/config/prelude.hpp>
+
+#include <cstddef>
+#include <stdexcept>
+#include <tuple>
+#include <type_traits>
+
+#include <bsoncxx/builder/basic/array.hpp>
+#include <bsoncxx/builder/basic/document.hpp>
+#include <bsoncxx/builder/basic/kvp.hpp>
+#include <bsoncxx/builder/basic/sub_array.hpp>
+#include <bsoncxx/builder/basic/sub_document.hpp>
+#include <bsoncxx/view_or_value.hpp>
+
+#include <bsoncxx/builder/stream/array.hpp>
+
+#define NVP(x) mongo_odm::makeNvp(&wrapBase::x, #x)
+
+#define ODM_MAKE_KEYS(Base, ...) \
+    using wrapBase = Base;       \
+    constexpr static auto fields = std::make_tuple(__VA_ARGS__);
+
+#define ODM_MAKE_KEYS_STORAGE(Base) constexpr decltype(Base::fields) Base::fields
+
+#define ODM_KEY(value) mongo_odm::hasCallIfFieldIsPresent<decltype(&value), &value>::call()
+
+namespace mongo_odm {
+MONGO_ODM_INLINE_NAMESPACE_BEGIN
+
+/**
+ * An object that represents a name-value pair of a member in an object.
+ * It is templated on the class of the member and its type.
+ */
+template <typename Base, typename T>
+struct Nvp {
+    /**
+     * Create a name-value pair from a member pointer and a name.
+     * @param  t    A pointer to the member
+     * @param  name The name of the member
+     */
+    constexpr Nvp(T Base::*t, const char *name) : t(t), name(name) {
```

Comment:
Makes sense to me, but unfortunately clang-format disagrees, I don't know if it's worth the effort of grappling with clang-format every time to keep the whitespace between `*` and `t`. Hope it's not too unclear as is.